File: app/src/main/java/me/eycia/Notifier.java
package me.eycia;

import java.util.ArrayList;

/**
 * Created by eycia on 16/5/12.
 */
public class Notifier {
    private final ArrayList<OnDataChangeListener> onDataChangeListeners = new ArrayList<>();

    public interface OnDataChangeListener {
        void OnDataChange();
    }

    public void addOnDataChangeListener(OnDataChangeListener listener) {
        synchronized (onDataChangeListeners) {
            if (!onDataChangeListeners.contains(listener)) {
                onDataChangeListeners.add(listener);
            }
        }
    }

    public void removeOnDataChangeListener(OnDataChangeListener listener) {
        synchronized (onDataChangeListeners) {
            if (onDataChangeListeners.contains(listener)) {
                onDataChangeListeners.remove(listener);
            }
        }
    }

    public void ChangeData() {
        for (OnDataChangeListener listener : onDataChangeListeners) {
            listener.OnDataChange();
        }
    }
}

File: app/src/main/java/me/eycia/Notifier.kt
package me.eycia

import java.util.*

/**
 * Created by eycia on 16/5/12.
 */
class Notifier {
    private val onDataChangeListeners = ArrayList<() -> Unit>()

    fun addOnDataChangeListener(listener: () -> Unit) {
        synchronized (onDataChangeListeners) {
            if (!onDataChangeListeners.contains(listener)) {
                onDataChangeListeners.add(listener)
            }
        }
    }

    fun removeOnDataChangeListener(listener: () -> Unit) {
        synchronized (onDataChangeListeners) {
            if (onDataChangeListeners.contains(listener)) {
                onDataChangeListeners.remove(listener)
            }
        }
    }

    fun ChangeData() {
        for (listener in onDataChangeListeners) {
            listener()
        }
    }
}
